services/sonar/sonar-coverage.tester.js

```diff
@@ -0,0 +1,22 @@
+'use strict'
+
+const t = (module.exports = require('../tester').createServiceTester())
+const { isIntegerPercentage } = require('../test-validators')
+
+t.create('Coverage')
+  .get(
+    '/http/sonar.petalslink.com/org.ow2.petals%3Apetals-se-ase/coverage.json'
+  )
+  .expectBadge({
+    label: 'coverage',
+    message: isIntegerPercentage,
+  })
+
+t.create('Coverage (legacy API supported)')
+  .get(
+    '/http/sonar.petalslink.com/org.ow2.petals%3Apetals-se-ase/coverage.json?version=4.2'
```

Comment:
Not sure quite where to leave a comment for it, but "version" is quite a loaded word. Do you think there is any ambiguity about what the `version` param means in this context?

i.e: `version` here refers to the version of SonarQube server I'm running as opposed to the version of my software (in the sense of a tag on a repository). Do you think it is worth changing the param name to make that more clear or noting that in the docs, or is that not an ambiguity we need to consider?

Happy to defer to your view on this as I'm not deeply familiar with SonarQube

Reply:
Also, the `version` param doesn't seem to be working at all on staging for any badges. If I call for example

https://shields-staging-pr-3189.herokuapp.com/sonar/http/sonar.petalslink.com/org.ow2.petals:petals-se-ase/coverage.svg?version=4.2 or https://shields-staging-pr-3189.herokuapp.com/sonar/https/sonarcloud.io/swellaby:azdo-shellcheck/quality_gate.svg?version=4.2

I get "invalid query parameter: version"

![](https://shields-staging-pr-3189.herokuapp.com/sonar/http/sonar.petalslink.com/org.ow2.petals:petals-se-ase/coverage.svg?version=4.2)
![](https://shields-staging-pr-3189.herokuapp.com/sonar/https/sonarcloud.io/swellaby:azdo-shellcheck/quality_gate.svg?version=4.2)

Reply:
🤔 not sure off hand but looking into it. it was working at one point (albeit admittedly another classic case of "works on my machine")

Reply:
> Do you think there is any ambiguity about what the version param means in this context?

I see what you mean. Honestly this is really more of a boolean determination anyway to answer the question: "is the SonarQube instance old (5.3 or earlier)". My personal preference would have been to have this as a boolean query param like `useLegacyApi`. 

However, I'm hesitant to change it this point unless we can find a clean way to maintain backwards compatibility without requiring a redirect every time (this could technically be done with a catch all redirector with a `transformQueryParam` value but I'm not keen on that approach)

Reply:
I believe this should be set now. I've renamed the query param to `sonarVersion` to be more explicit, and I fixed the previous issue with the query param (was related to the query param schema validation):

https://shields-staging-pr-3189.herokuapp.com/sonar/http/sonar.petalslink.com/org.ow2.petals:petals-se-ase/coverage.svg?sonarVersion=4.2
![](https://shields-staging-pr-3189.herokuapp.com/sonar/http/sonar.petalslink.com/org.ow2.petals:petals-se-ase/coverage.svg?sonarVersion=4.2)

https://shields-staging-pr-3189.herokuapp.com/sonar/http/sonar.petalslink.com/org.ow2.petals:petals-se-ase/quality_gate.svg?sonarVersion=5.4
![](https://shields-staging-pr-3189.herokuapp.com/sonar/http/sonar.petalslink.com/org.ow2.petals:petals-se-ase/alert_status.svg?sonarVersion=5.4)

https://shields-staging-pr-3189.herokuapp.com/sonar/https/sonarcloud.io/swellaby:azdo-shellcheck/quality_gate.svg?version=7.7
![](https://shields-staging-pr-3189.herokuapp.com/sonar/https/sonarcloud.io/swellaby:azdo-shellcheck/quality_gate.svg?version=7.7)

Reply:
Nice - this is looking good now. With regard to treating it as a boolean for current/legacy, I think you're right to explicitly pass a version number. Thinking ahead, a future versions might introduce more API changes so we could in principle need to support more than 2 formats.